R/module_teal_lockfile.R

```diff
@@ -0,0 +1,219 @@
+#'     `teal.lockfile.path` option. Automatic lockfile computation is skipped in such case.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `teal.lockfile.path` is not set, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application's environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return `NULL`
+#'
+#' @name module_teal_lockfile
+#' @rdname module_teal_lockfile
+#'
+#' @keywords internal
+NULL
+
+#' @rdname module_teal_lockfile
+ui_teal_lockfile <- function(id) {
+  ns <- NS(id)
+  shiny::tagList(
+    tags$span("", id = ns("lockFileStatus")),
+    shinyjs::disabled(downloadLink(ns("lockFileLink"), "Download lockfile"))
+  )
+}
+
+#' @rdname module_teal_lockfile
+srv_teal_lockfile <- function(id) {
+  moduleServer(id, function(input, output, session) {
+    logger::log_debug("Initialize srv_teal_lockfile.")
+    enable_lockfile_download <- function() {
+      shinyjs::html("lockFileStatus", "Application lockfile ready.")
+      shinyjs::hide("lockFileStatus", anim = TRUE)
+      shinyjs::enable("lockFileLink")
+      output$lockFileLink <- shiny::downloadHandler(
+        filename = function() {
+          "renv.lock"
+        },
+        content = function(file) {
+          file.copy(lockfile_path, file)
+          file
+        },
+        contentType = "application/json"
+      )
+    }
+    disable_lockfile_download <- function() {
+      warning("Lockfile creation failed.", call. = FALSE)
+      shinyjs::html("lockFileStatus", "Lockfile creation failed.")
+      shinyjs::disable("lockFileLink")
+    }
+
+    shiny::onStop(function() {
+      if (file.exists(lockfile_path) && !shiny::isRunning()) {
+        logger::log_debug("Removing lockfile after shutting down the app")
+        file.remove(lockfile_path)
+      }
+    })
+
+    lockfile_path <- "teal_app.lock"
+    mode <- getOption("teal.lockfile.mode", default = "")
+    user_lockfile_path <- getOption("teal.lockfile.path", default = "")
+
+    if (!(mode %in% c("auto", "true", "false", "user"))) {
```

Comment:
`%in%` has precendence before `!`. 
Also mode should be problably "true" -> "enable", "false" -> "disable" as "true" and "false" are boolean values and having something else is weird in this context.
```suggestion
    if (!mode %in% c("auto", "true", "false", "user")) {
```

Reply:
how about "on"/"off"?

Reply:
@gogonzo I know that

```r
!x %in% y
```

is the same as

```r
!(x %in% y)
```

but it always makes me confused when there are no parenthesis as it's not the same for `&` operator. So I better put parenthesis so the order is clear

Reply:
When it comes to the name, I'd vote for `enabled/disabled` or `enable/disable`

Reply:
> it always makes me confused when there are no parenthesis

I recommend highly `is.element` 😉